### PR TITLE
[DTensor] single dim fix inplace op expansion

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172278
* #172479
* #172478
* #172282
* __->__ #172477
* #172420

This enables the inplace filtering logic that skips strategies with
incompatible input placements.

Previously, inplace ops were able to generate expanded strategies
incompatible with the current input placements, which aren't allowed to
be modified by redistribution.

### CODE REVIEW

File: docs/source/distributed.tensor.md [hunk 1]
```diff
@@ -111,6 +111,59 @@ DTensor supports the following types of {class}`Placement` on each {class}`Devic
...
+### Partial Ordering Invariant
+
+When a DTensor has multiple {class}`Partial` placements with **different reduce operations**
+(e.g., `Partial("sum")` on one mesh dimension and `Partial("max")` on another), the order in
+which these reductions are applied is semantically significant. This is called the
+**partial ordering invariant**.
+
+**Left-to-Right Evaluation Order**
+
+DTensor evaluates Partial placements in **left-to-right order** (i.e., mesh dimension 0 first,
```
Comment (zpcore) [inline docs/source/distributed.tensor.md]: I checked the redistribution code, **left-to-right order** is not completely correct.
The Partial-> Replicate can be right-to-left order:
https://github.com/pytorch/pytorch/blob/ee562d97dfdc683ade44a4883741c6b4005ecef3/torch/distributed/tensor/_redistribute.py#L749-L757
or left-to-right:
https://github.com/pytorch/pytorch/blob/ee562d97dfdc683ade44a4883741c6b4005ecef3/torch/distributed/tensor/_redistribute.py#L765-L773
But in general, it should be either one.

Comment (zpcore) [reply docs/source/distributed.tensor.md]: As long as we have Shard in src placement, we can trigger right-to-left:
```python
from torch.utils._debug_mode import DebugMode
class DistributeWithPartialTest(DTensorTestBase):
    @property
    def world_size(self) -> int:
        return 8

    def _extract_redistribute_trace_from_debug_mode(self, s: str) -> str:
        import re

        match = re.search(r"trace:\s*(.*)\)", s)
        if match:
            trace_str = match.group(1)
            return trace_str
        else:
            return ""

    @with_comms
    def test_tmp(self):
        mesh = init_device_mesh(self.device_type, (2, 2, 2))
        input_data = torch.randn((8, 8, 8), device=self.device_type)
        dt = DTensor.from_local(input_data, mesh, (Partial('sum'), Shard(0), Partial('max')))
        with DebugMode(record_torchfunction=False) as debug_mode:
            dt2 = dt.redistribute(mesh, (Replicate(), Replicate(), Replicate()))
        trace_str = self._extract_redistribute_trace_from_debug_mode(debug_mode.debug_string())
        print(trace_str)
```
The redistribution path is: `P(sum)S(0)P(max)->P(sum)S(0)R->P(sum)RR->RRR`. This made me think that there is a bug with the greedy redistribution algorithm when handling Partial with variant. We can't arbitrarily switch between left-to-right or right-to-left order. I think right-to-left should be the correct order to handle Partial->Replicate.

File: torch/distributed/tensor/_redistribute.py [hunk 2]
```diff
@@ -713,6 +761,68 @@ def generate_greedy_transform_infos(
...
+        #
+        # The correct sequence is:
+        # 1. Reduce all Partials to Replicate first
+        # 2. Then re-partition back to the target Partial placements
+        src_partial_ops = {}  # mesh_dim -> reduce_op
+        for mesh_dim, placement in enumerate(current_placements):
+            if isinstance(placement, Partial):
+                src_partial_ops[mesh_dim] = placement.reduce_op
+
+        if len(src_partial_ops) > 1:
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py]: Aha, I see. I left the comment in README file too early. This PR is enforcing the left-to-right order when there exists multiple Partial variants.

File: torch/distributed/tensor/_redistribute.py [hunk 3]
```diff
@@ -713,6 +761,68 @@ def generate_greedy_transform_infos(
...
+                dst_partial_dims = set()
+                for mesh_dim, placement in enumerate(target_placements):
+                    if isinstance(placement, Partial):
+                        dst_partial_dims.add(mesh_dim)
+
+                src_partial_dims = set(src_partial_ops.keys())
+
+                # If some partials are being reduced while others are kept,
+                # we need to reduce ALL partials first, then re-partition
+                if dst_partial_dims and dst_partial_dims != src_partial_dims:
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py]: If the condition `dst_partial_dims and dst_partial_dims != src_partial_dims` doesn't hold, then we will use the default greedy code to handle partial, which can either be left-to-right or reverse. We also need to enforce that order.

File: docs/source/distributed.tensor.md (hunk 1 above)
Comment (wconstab) [reply docs/source/distributed.tensor.md]: That sounds good to me. However don't move too fast on it, we are discussing just banning mixed partial types existing in the same placement and then we don't need to define an
order

Comment (wconstab) [reply docs/source/distributed.tensor.md]: oh, this PR was not even supposed to include the changes to partial order. it was a rebasing mistake. I will remove that code.  I have another PR up now anyway that partially bans mixed partials.  If people are OK with that direction, i'll extend that and ensure we don't allow mixed partials at all parts of the DTensor stack.  then this ordering stuff becomes unnecessary.

Comment (zpcore) [reply docs/source/distributed.tensor.md]: Yes, "banning mixed Partial" sounds good to me!

**Review by zpcore — APPROVED** [review]
LGTM!
